doc/SEMANTICS.md

```diff
@@ -63,9 +63,14 @@ Mountpoint has limited support for other file and directory metadata, including
 
 ## Consistency and concurrency
 
-Amazon S3 provides [strong read-after-write consistency](https://docs.aws.amazon.com/AmazonS3/latest/userguide/Welcome.html#ConsistencyModel) for PUT and DELETE requests of objects in your S3 bucket. Mountpoint provides strong read-after-write consistency for file writes, directory listing operations, and new object creation. For example, if you create a new object using another S3 client, it will be immediately accessible with Mountpoint. Mountpoint also ensures that new file uploads to a single key are atomic. If you modify an existing object in your bucket with another client while also reading that object through Mountpoint, the reads will return either the old data or the new data, but never partial or corrupt data. To guarantee your reads see the newest object data, you can re-open the file after modifying the object.
+Amazon S3 provides [strong read-after-write consistency](https://docs.aws.amazon.com/AmazonS3/latest/userguide/Welcome.html#ConsistencyModel) for PUT and DELETE requests of objects in your S3 bucket.
+By default, Mountpoint provides strong read-after-write consistency for file writes, directory listing operations, and new object creation. For example, if you create a new object using another S3 client, it will be immediately accessible with Mountpoint. Mountpoint also ensures that new file uploads to a single key are atomic. If you modify an existing object in your bucket with another client while also reading that object through Mountpoint, the reads will return either the old data or the new data, but never partial or corrupt data. To guarantee your reads see the newest object data, you can re-open the file after modifying the object.
 
-However, Mountpoint may return stale metadata for an existing object within 1 second of the object being modified or deleted in your S3 bucket by another client. This occurs only if the object was accessed through Mountpoint immediately before being modified or deleted in your S3 bucket. The stale metadata will only be visible through metadata operations such as `stat` on individual files. Directory listings will never be stale and always reflect the current metadata. These cases do not apply to newly created objects, which are always immediately visible through Mountpoint. Stale metadata can be refreshed by either opening the file or listing its parent directory.
+Mountpoint also offers metadata and object content caching which can be enabled using CLI flags: see the [caching section of the configuration document](./CONFIGURATION.md#caching) for more information.
+When opting into caching, the consistency model is relaxed and you may see stale entries until they have expired.
+Stale entries may live as long as the cache's metadata time-to-live (TTL).
```

Comment:
Is there an escape hatch to guarantee consistency? (`O_DIRECT` for data, anything for metadata?)

Reply:
`O_DIRECT` should definitely bypass the cached data. I think it's an open question whether it should _also_ bypass cached metadata and force an ETag lookup. Strictly speaking, that's not what `O_DIRECT` is documented to do, but that's because in a real file system `lookup` is supposed to do all the work of validating metadata. But here, if we're bypassing the cached data we also need a way to find out what the latest ETag is, so forcing a re-lookup is probably the right thing to do?

Reply:
I'd challenge if we should bypass cached data. `O_DIRECT` is a hint to the Kernel to avoid the page cache and read directly from storage. As I understand it, it can be ignored and even if we ignore it, users would still benefit from bypassing the page cache I assume.

I would assume that we want to refresh the ETag, so it should trigger a lookup that will refresh that. If it hasn't changed, then we know that the data on our disk cache is already up to date with S3 at the time of opening. I don't see benefit to the user to bypass that cache?

Reply:
`O_DIRECT` isn't specific to the page cache, the hint is "Try to minimize cache effects of the I/O to and from this file". In practice people use it to mean "hit the storage device directly". It's true that this is a case where we can detect whether the cache is valid or not, but we should probably just do what people expect even if it's slower.

Reply:
I've updated this (although implementation is yet to be completed!).